process/task/terminal: refactor escaping/quoting

marechal-p opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md
-->

#### What it does

Add utility functions in `@theia/process/lib/common` to escape common
shells' arguments. Refactored terminal processes to not handle shell
escaping anymore, it is the caller's responsability to provide the
escaped spawn options.

Escaping is now done for DAP's `runInTerminal` requests.

Most of the design is taken from VS Code, with minor tweaks.

Tried to test as much as possible since escaping rules can get complicated in corner cases.

Fixes https://github.com/eclipse-theia/theia/issues/6684

#### How to test

1. Build and then run the test suites (`@theia/process`, `@theia/task`, and `@theia/debug`).
2. Test debugging a Python file:
    1. Download [ms-python.vsix](https://github.com/Microsoft/vscode-python/releases/download/2019.5.18875/ms-python-release.vsix)
    2. Setup debug config to run opened Python files.
    3. Try to debug a Python file with space in its filename, it should work with this patch but not without.
3. Test different `shell` tasks, using quoted strings (see [VS Code Custom Tasks](https://code.visualstudio.com/docs/editor/tasks#_custom-tasks)).

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

marechal-p commented on packages/process/src/node/terminal-process.spec.ts:
```diff
@@ -103,75 +103,75 @@ describe('TerminalProcess', function (): void {
-    });
-
-}
+// /**
```
Need to fix that, forgot.

akosyakov commented on packages/debug/src/browser/debug-session.tsx:
```diff
@@ -379,8 +383,13 @@ export class DebugSession implements CompositeTreeElement {
+        //       into a function.
+        // Parallelize promise resolution by requesting both now:
+        const { processId, processInfo } = terminal;
+        terminal.sendText(this.runInTerminalService.prepareCommandLine(await processInfo, { cwd, args, env }) + '\n');
```
Could we encapsulate it in `terminal` instead of  `RunInTerminalService`?

i.e. a client will simply call `await terminal.sendCommand({ cwd, args, env });`?

marechal-p replied:
:thinking: while I really don't like writing commands in place of a user, we can do that since the code is there now anyway. Will do.

akosyakov commented on packages/debug/src/browser/run-in-terminal.ts:
```diff
@@ -0,0 +1,145 @@
+}
+
+@injectable()
+export class RunInTerminalService {
```
EIther prefix with `Debug` or move to `@theia/terminal`, please see https://github.com/eclipse-theia/theia/wiki/Coding-Guidelines#unique-names

marechal-p replied:
RunInTerminal felt like a debug-specific concept, but you are right. Will move.

akosyakov commented on packages/terminal/src/browser/base/terminal-widget.ts:
```diff
@@ -24,6 +25,8 @@ export abstract class TerminalWidget extends BaseWidget {
 
     abstract processId: Promise<number>;
 
+    abstract processInfo?: Promise<TerminalProcessInfo>;
```
you can do `:`, there won't be other implementations of `TerminalWidget`

at some point we will delete `TerminalWidget` interface and rename `TerminalWidgetImpl` to avoid bolierplate

akosyakov commented on packages/task/src/node/task-server.slow-spec.ts:
```diff
@@ -126,37 +126,26 @@ describe('Task server / back-end', function (): void {
     });
 
-    it('task is executed successfully with cwd as a file URI', async function (): Promise<void> {
+    it.only('task is executed successfully with cwd as a file URI', async function (): Promise<void> {
```
`it.only` leftover?

akosyakov commented on packages/process/src/common/shell-quoting.ts:
```diff
@@ -0,0 +1,194 @@
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
```
please add a link to the original code

akosyakov commented on packages/debug/src/browser/debug-session-contribution.ts:
```diff
@@ -134,7 +138,11 @@ export class DefaultDebugSessionFactory implements DebugSessionFactory {
+            this.fileSystem,
+        );
+        // Property injections
+        debugSession['runInTerminalService'] = this.runInTerminalService;
```
:D

elaihau commented:
In theia with this change:
![Peek 2020-01-07 00-12](https://user-images.githubusercontent.com/37082801/71870048-6e549300-30e2-11ea-93e5-516ef01ad14d.gif)



in vscode
![Peek 2020-01-07 00-13](https://user-images.githubusercontent.com/37082801/71870092-98a65080-30e2-11ea-8078-0c282f3de8f0.gif)


Task I tested with 
```
      {
        "type": "shell",
        "command": "node",
        "args": [
          "-e",
          "\"setTimeout(()=>console.log(\\\"1 + 2\\\"),1000)\""
        ],
        "label": "test shell 211"
      }
```

marechal-p commented:
Thanks for the test @elaihau ! I was missing some task related tests, added them back and fixed the issue you reported :)

Although, the specific example you are using is annoying me in some way, because in my opinion it illustrates a behavior that I don't like from VS Code: Essentially, they battle to see if a user already escaped part of an argument, see [related code](https://github.com/microsoft/vscode/blob/d363b988e1e58cf49963841c498681cdc6cb55a3/src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts#L1101-L1127).

From what I understand, when dealing with user defined shell task arguments, I see 2 scenarios:

1. User just writes the fully escaped command line in the `command` field, this is executed as is. 
 
    This is the simplest case, since we can assume that it is the user's responsibility to properly escape arguments.

2. User write an executable path in the `command` field, and provides un-escaped arguments. 
 
    Given the tests I have written, our escaping functions should handle about anything. This means that the user should not care about how the arguments will be passed to the shell, since this is now Theia's job to properly escape everything. You can still somewhat control how Theia will escape your argument by using the `{ value: 'a b', quoting: 'escape' }` notation to tell it what strategy to use, making it Theia's responsibility.

    But VS Code seems to assume that people might try to escape their arguments themselves. To me this is more confusing than anything. What's more, they try to detect things such as partially escaped sequences: `/path/with/"white space"/meh`. This -to me- is an issue when the quotation mechanism is weak, and you lay the responsibility on the user to do it properly. Then if VS Code detects that quotation is still needed because some white space was missed, it will simply overlay quotes around the whole thing. So `/path/with/"white space"/and other white space/meh` becomes `'/path/with/"white space"/and other white space/meh'`, so now the inner quotes are useless, since VS Code might quote in a way that will make them break the path.

    Example 1:

    ```json
    {
        "type": "shell",
        "command": "node",
        "args": [
            "-e",
            "setTimeout(()=>console.log('1 + 2'), 1000)"
        ],
        "label": "test shell 211"
    }
    ```

    Breaks, because VS Code sees the whitespace in front of `1000` and adds strong quotes, without caring about the content of what is being quoted, the single quotes already in place interfere and the command fails. Theia's implementation wouldn't fail.

   Example 2:

    ```json
    {
        "type": "shell",
        "command": "node",
        "args": [
            "-e",
            "setTimeout(()=>console.log(process.argv), 1000)",
            "/path/with/\"white space\"/and other white space/meh"
        ],
        "label": "test shell 322"
    }
    ```

    This one also doesn't pass the right thing to the shell because the single quotes interfere with the quoting added by VS Code. Using double quote wouldn't fix much either since it will embed them into the path, which is not what you want.

    Example 3:

    ```json
    {
        "type": "shell",
        "command": "node",
        "args": [
            "-e",
            "setTimeout(()=>console.log(1+2),1000)"
        ],
        "label": "test shell 433"
    }
    ```

    This one is even better, because even without white space, the command will fail if passed as is to bash. But since VS Code only looks at white space, this is passed un-escaped and fails. The best would be to simply escape everything properly, no questions asked. But in this case I slightly modified VS Code's trigger characters to fix that issue now that I can.

     Conclusion: It is simplier to assume that arguments are un-escaped and that we should do it ourselves, rather than this mixed mode where it is confusing as to how should we quote things inside a JSON string, which we also must take care of escaping for (escape quotes for JSON, to escape stuff for shell X).

With all of that said, I still tried to mimic what VS Code is doing, without doing it the same way either. The quotation mechanism should be more stable and robust, and user shouldn't care about escaping a thing, unless it is running in an unknown shell for which Theia doesn't know how to escape. But that's kinda asking for trouble, and practically it shouldn't happen.

Now if you have a task command complex enough that you need crazy escaping, that Theia would natively handle but not VS Code, the solution is simply to refactor the command into a script and write a simple task to trigger said script, it will avoid a lot of trouble for everyone ahah.

marechal-p commented:
I would even go as far as adding a property for shell tasks named `verbatimArguments: boolean`, which would prevent any automatic quoting, so that white space or not, the user takes responsibility in regard to quoting. Even though in this case it might be 10x easier to not use arguments and rather craft a single command line, which will already be passed verbatim.

akosyakov commented on packages/debug/package.json:
```diff
@@ -67,7 +67,8 @@
   "devDependencies": {
-    "@theia/ext-scripts": "^0.14.0"
+    "@theia/ext-scripts": "^0.14.0",
+    "colors": "^1.4.0"
```
left overs? Where is it used?

Please keep dev dependencies in root package to ensure that we don't use many different versions in the repo.

marechal-p replied:
It used for tests, will move at the root.

akosyakov commented on packages/debug/src/browser/debug-session.tsx:
```diff
@@ -64,6 +65,9 @@ export class DebugSession implements CompositeTreeElement {
     protected readonly toDispose = new DisposableCollection();
 
+    // @inject(RunInTerminalUtility)
+    protected runInTerminalUtility: RunInTerminalUtility;
```
unused, please remove

akosyakov commented on packages/debug/src/browser/debug-session-contribution.ts:
```diff
@@ -114,6 +115,8 @@ export class DefaultDebugSessionFactory implements DebugSessionFactory {
     @inject(FileSystem)
     protected readonly fileSystem: FileSystem;
+    @inject(RunInTerminalUtility)
+    protected readonly runInTerminalUtility: RunInTerminalUtility;
```
looks like this file can be reverted

marechal-p replied:
Yes, I pushed my changes mid-edition yesterday evening, I was planning on finishing today :)

akosyakov commented on packages/process/src/common/run-in-terminal.ts:
```diff
@@ -0,0 +1,152 @@
+ * Create command lines ready to be sent to a shell's stdin for evaluation.
+ */
+@injectable()
+export class RunInTerminalUtility {
```
Also looking at the class it is not clear what it has to do with terminals? This class does not run anything in terminal. Maybe `ProcessCommandBuilder.build` instead? then it is relevant to the process and clear what it does. Or can we just inline it in `TerminalWidget`? Where else it supposed to be used?

marechal-p replied:
"Run In Terminal" as in running things in a terminal? And the comment explaining what it does? Will rename to "ShellCommandBuilder".

akosyakov commented on packages/terminal/src/browser/base/terminal-widget.ts:
```diff
@@ -16,6 +16,8 @@
 import { Event } from '@theia/core';
 import { BaseWidget } from '@theia/core/lib/browser';
+import { TerminalProcessInfo } from '../../common/base-terminal-protocol';
+import { CommandLineOptions } from '@theia/process/libcommon/run-in-terminal';
```
`lib/common`

marechal-p replied:
Somehow this is the works of TypeScript's auto-import, idk why it keeps doing that???

akosyakov commented on packages/terminal/src/browser/base/terminal-widget.ts:
```diff
@@ -36,6 +40,13 @@ export abstract class TerminalWidget extends BaseWidget {
+     *
+     * @param commandOptions - command, arguments and cwd.
+     */
+    abstract buildAndExecuteCommand(commandOptions: CommandLineOptions): Promise<void>;
```
`build` bit seems to be implementation detail, the real intention is `executeCommand`

akosyakov commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -423,6 +432,10 @@ export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget
         }
     }
 
+    async buildAndExecuteCommand(commandOptions: CommandLineOptions): Promise<void> {
```
async?

marechal-p replied:
Should it not be?

marechal-p commented:
> We can do as we like, but we should support whatever VS Code allows as minimum.

Indeed, but to me it seems a bit like bad UX from VS Code, having to think about 3 escaping stages (JSON, VS Code, Shell) instead of the minimum of 2 (JSON, Shell).

akosyakov commented on package.json:
```diff
@@ -22,6 +22,7 @@
     "@types/webdriverio": "^4.7.0",
     "chai": "^4.1.0",
     "chai-string": "^1.4.0",
+    "colors": "^1.4.0",
```
@marechal-p Why is it needed for this PR? sorry, I could not figure it out

marechal-p replied:
https://github.com/eclipse-theia/theia/blob/727b2762945048ea62ca224bfec6882cebecca90/packages/process/src/common/shell-command-builder.spec.ts#L31

akosyakov replied:
ok

akosyakov commented on packages/process/src/node/terminal-process.ts:
```diff
@@ -187,7 +53,7 @@ export class TerminalProcess extends Process {
         @inject(MultiRingBuffer) protected readonly ringBuffer: MultiRingBuffer,
         @inject(ILogger) @named('process') logger: ILogger
     ) {
-        super(processManager, logger, ProcessType.Terminal, TerminalProcess.resolveExecOptions(options));
```
Could it break some existing users? Or escaping done in another place for existing clients?

marechal-p replied:
Escaping is done at the terminal's caller level. It simplifies the role of the `TerminalProcess` class, since its original purpose was simply to start arbitrary processes inside a pty, and not necessarily inside a shell.

marechal-p replied:
Depending on how people used that class, it might be breaking, but I think it is for the best.

akosyakov replied:
a record in the changelog explaining what clients should do instead?

akosyakov commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -423,6 +432,10 @@ export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget
     }
 
+    async executeCommand(commandOptions: CommandLineOptions): Promise<void> {
+        this.sendText(this.runInTerminalUtility.buildCommand(await this.processInfo, commandOptions) + '\n');
```
Why is it async? We don't guarantee anything here for a client by awaiting. It should be `void`. `sendText` is `void` as well indicating that it does not return then text is processed.

marechal-p replied:
I don't know if `await this.processInfo` could fail, people could want to make sure that the call did not fail?

akosyakov replied:
ok, please add js-doc what resolve means, more important that it does not that command completed, for that they would need some other APIs

marechal-p commented:
Having little issues with Windows, will fix.

vince-fugnitto commented on package.json:
```diff
@@ -22,6 +22,7 @@
     "@types/webdriverio": "^4.7.0",
     "chai": "^4.1.0",
     "chai-string": "^1.4.0",
+    "colors": "^1.4.0",
```
What is the motivation behind including `colors` in the pull-request?
Is it solely used as a visual aid during debugging?

Please register a `devDependency` CQ for the following if it is not already covered:
https://github.com/eclipse-theia/theia/wiki/Registering-CQs#case-new-or-updated-testbuild-dependency-eg-devdependencies

marechal-p replied:
It is used for debugging, it really helps making sense of the different streams produced during the test execution. Will check it out following your link.

vince-fugnitto commented on packages/process/src/common/shell-command-builder.spec.ts:
```diff
@@ -0,0 +1,444 @@
+/********************************************************************************
+ * Copyright (C) 2019 Ericsson and others.
```
2020, and as well in other files.

vince-fugnitto commented on packages/process/src/common/shell-command-builder.spec.ts:
```diff
@@ -0,0 +1,444 @@
+            ? convertWindowsPath(testResources)
+            : testResources;
+
+        // When running powershell, it seems like good mesure to send `\n` twice...
```
typo - `mesure` > `measure`.

vince-fugnitto commented on packages/process/src/common/shell-command-builder.ts:
```diff
@@ -0,0 +1,152 @@
+ ********************************************************************************/
+
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
```
Please file the appropriate CQ for the copied/inspired code.

vince-fugnitto commented on packages/process/src/common/shell-quoting.spec.ts:
```diff
@@ -0,0 +1,176 @@
+                ]
+            },
+        },
+        // https://ss64.com/nt/syntax-esc.html ?
```
I don't understand the following comment or the need for `?`.

marechal-p replied:
These are simply references, might be handy to understand the tests. Removed the "?".

vince-fugnitto commented on packages/process/src/common/shell-quoting.spec.ts:
```diff
@@ -0,0 +1,176 @@
+        }
+    } as const;
+
+    // iter through each runtime (bash/cmd/powershell)
```
[minor] these comments, and the ones below are inconsistent with the rest of the file, mainly the format of capitalized case ended in punctuation.

vince-fugnitto commented on packages/process/src/common/shell-quoting.ts:
```diff
@@ -0,0 +1,220 @@
+// #region vscode
+
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
```
Please file the appropriate CQ if not already done so.

vince-fugnitto commented on packages/process/src/node/terminal-process.ts:
```diff
@@ -17,164 +17,30 @@
 import { signame } from './utils';
 import { Writable } from 'stream';
 
-export type QuotingType = 'escaped' | 'strong' | 'weak';
```
I think all these deletions should be documented at least in the changelog, since they will break anyone who references them.

akosyakov commented on .travis.yml:
```diff
@@ -1,95 +1,96 @@
   global:
     - CXX=g++-4.8
     - NODE_OPTIONS="--max_old_space_size=4096"
+    - THEIA_TEST_DEBUG=1
```
are changes to this file necessary? seems to be a lot of noise and what is `THEIA_TEST_DEBUG`?

marechal-p replied:
This is a desperate attempt at understanding why travis fails, this is temporary.

akosyakov replied:
ok, please don't forget to revert, i will skip looking at the PR till the build is green to avoid noise from me 😄

akosyakov commented on packages/process/src/common/shell-command-builder.spec.ts:
```diff
@@ -0,0 +1,450 @@
+}
+
+const isWindows = process.platform === 'win32';
+const _debug = !!process.env['THEIA_TEST_DEBUG'];
```
I don't like a new env variable which is used only by one test. Could we get rid of it and just always log necessary information but not more as we do for other tests?

marechal-p replied:
I really prefer being able to do extra logging in some way. The information being logged when this flag is enabled is useful when trying to understand what happens inside the shells, it is only really useful when developing the tests. Environment variables cost nothing, I don't plan on making this one standard, and anyone willing to investigate a test can quickly understand what he has to do with it: Set this var and you will get more info while the test is running.

akosyakov replied:
I usually rerun and debug only failing tests to see the reason, although agree that sometimes extra logging helps.

akosyakov commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -423,6 +432,19 @@ export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget
+     *
+     * Supported shells: `bash`, `cmd.exe`, `wsl.exe`, `pwsh/powershell.exe`
+     */
+    async executeCommand(commandOptions: CommandLineOptions): Promise<void> {
```
docs should be aligned in  `TerminalWidget` and `TerminalWidgetImpl`

marechal-p replied:
ah true, will fix.

marechal-p commented:
I just want Travis to run at this point. I might just skip whatever doesn't want to work (bash.exe on Windows seems to run but stdio is dead :/)

akosyakov commented:
Any plans to finish it?

marechal-p commented:
Pending: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=21515

vince-fugnitto commented:
@marechal-p do you mind rebasing the pull-request (resolve the conflicts), I can take another high-level look at the pull-request :)

marechal-p commented:
I will now have to fix the new issues on Windows that appeared after the rebase, sigh.

CQ still pending too, so it's fine I guess.

akosyakov commented:
@marechal-p CQ has checkin tag, it means we can merge

tom-shan commented:
Looking forword to the landing of this PR which will fix the debug failure when using existing `Python Debug Console` of vscode python plugin in theia.

marechal-p commented:
@tom-shan I'm having issues with escaping for `cmd.exe` for tasks, I will take another crack at it Wednesday, once I get to work on a Windows machine. This shell alone makes everything more difficult, but oh well.

marechal-p commented:
I think I managed to tackle most of the issues I was having. Hopefully the CI will be green now.

This PR is finally re-ready for review. Sorry for the delay, it definitely was a struggle.

The code might not be perfect, but I'll leave this up to reviewers.

akosyakov commented on packages/core/src/node/backend-application.ts:
```diff
@@ -101,6 +101,8 @@ export class BackendApplication {
     @inject(ApplicationPackage)
     protected readonly applicationPackage: ApplicationPackage;
 
+    private _performanceObserver: PerformanceObserver;
```
`readonly`

marechal-p commented on packages/process/src/common/shell-command-builder.ts:
```diff
@@ -0,0 +1,157 @@
+
+export interface ProcessInfo {
+    executable: string
+    arguments: string[]
```
Here: https://github.com/eclipse-theia/theia/pull/6836/files#diff-29ae8bb3747051f34ef7d32716b705e6R511

kittaakos commented:
@marechal-p, I have noticed test failures in `@theia/task`:
 - There are [test failures](https://github.com/eclipse-theia/theia/issues/7838#issue-620760829) on macOS (running them locally, we have verified on multiple OS X),
 - There are [test failures on Linux too](https://github.com/eclipse-theia/theia/issues/7838#issuecomment-632650809) (in Gitpod).

I was looking into the history of `task-server.slow-spec.ts`, and it seems the last change was with this PR. I have switched back to ce398728 which is the parent commit of your PR. Without your change, the tests are green. Can this PR break tests? Do the tests pass on your side, not on the CI? Thanks!

CC: @akosyakov

marechal-p commented:
@kittaakos 

> Do the tests pass on your side, not on the CI?

Hmm this PR was merged early April, since then the CI was green on all 3 platforms. @vince-fugnitto  tried to run master's task tests on his own OS X and the only failure seemed to be related to nvm behaving weird and killing a terminal (1 failure only).

Is it possible that one of our dependencies updated? (not necessarily npm-related)

Something new in the environments where the failure happens?

note; I just tried on my Windows machine and the tests are passing.

kittaakos commented:
> note: I just tried on my Windows machine and the tests are passing.

Yes, I know. BTW, some tests are skipped on Windows.

> Hmm this PR was merged early April, since then the CI was green on all 3 platforms.

I know. It is strange, I agree.

> s it possible that one of our dependencies updated? (not necessarily npm-related)
> 
> Something new in the environments where the failure happens?

Well, I do not know, but [I can reproduce the test failure with this change and it works without](https://github.com/eclipse-theia/theia/issues/7838#issuecomment-633601309), so I do not think this is a non-npm dependency issue.

marechal-p commented:
Just tried on an Ubuntu 18.04 and no failure either.

I cannot efficiently troubleshoot your issue if I cannot reproduce though... Can you try to find the cause of the failures caused by this patch in your environment?

kittaakos commented:
> Can you try to find the cause of the failures caused by this patch in your environment?

Where did you get [these](https://github.com/eclipse-theia/theia/pull/6836/files#diff-01e0125ed48100ea02f3c69f5d88d2a6R175-R190), @marechal-p? For example, we did not have this code, or similar before, my command is `/bin/zsh`, it won't match any of the regexpes, is that expected? `zsh` is the default shell in macOS (10.15.+) This explains why the tests pass on the OS X CI but not locally.

kittaakos commented:
OK, I got it. Seems like we are already outdated compared to VS Code:
https://github.com/microsoft/vscode/blob/7e407b547294849e9a22b16e5aea28a56f88524c/src/vs/workbench/contrib/debug/node/terminals.ts#L78-L197

marechal-p commented:
Maybe just a matter of picking better defaults than what we currently have, see VS Code:

```ts
	} else if (env.isWindows) {
		shellType = ShellType.cmd; // pick a good default for Windows
	} else {
		shellType = ShellType.bash;	// pick a good default for anything else
	}
```

Currently there's no default being used.
